### DISCUSSION TITLE
Saving entries field doesn't work

### DISCUSSION THREAD

**Discussion by JulieVelghe** [craftcms/cms#15263] [Q&A] [answered]
Hi all,
I have a question about some custom development I'm doing.
I have a registration form that has an elements integration to create a user. The user has an entry field representing a chosen field of study. The data for that field is on the main site.
The form is on a subsite, and it has a field where the user can select an entry from the main site (being a chosen field of study).

On the initial creating of the user, this field is not getting mapped and remains empty.

I have another form, to modify the user data. I used a custom module, where I checked in the event_after_submission, which field of study was chosen on that form and update the user data. That code works well:

```
                    if ($form->handle == "formUserData"){
                        $primarySiteId = Craft::$app->sites->primarySite->id;

                        //get the field userCourse and save it to the user's field userCourse
                        $submissionUserCourse = $submission->getFieldValue('userCourse')->siteId($primarySiteId)->one();
                        $submissionUserCourseId = $submissionUserCourse->id;
                        $submissionUserId = $submission->getFieldValue('userProfileId');

                        //find user
                        $user =  \craft\elements\User::find()->id($submissionUserId)->one();
                        $user->setFieldValue('userCourse', $submissionUserCourseId);
                        Craft::$app->elements->saveElement($user);
                    }
```

I have created another function now to try to map the user course field:

```
        Event::on(Elements::class, Elements::EVENT_AFTER_SAVE_ELEMENT, function(ElementEvent $event) {
            if ($event->element instanceof \craft\elements\User && !ElementHelper::isDraftOrRevision($event->element)) {

                $user = $event->element;
                $isNewUser = $event->isNew;

                if ($user->can('Admin')) {
                    // leave admins alone
                    return;
                }

                if ($isNewUser && $user->isInGroup('creators')) {
                    // This is a new user
                    $tempcourseid = $_POST['fields']['userCourse'][0];

                    $primarySiteId = Craft::$app->sites->primarySite->id;

                    $submissionUserCourse = \craft\elements\Entry::find()->section('courses')->id($tempcourseid)->siteId($primarySiteId)->one();

                    $submissionUserCourseId = $submissionUserCourse->id;

                    Craft::info('User is new, usercourseid is '.$submissionUserCourseId, __METHOD__);

                    //find user
                    $userel =  \craft\elements\User::find()->id($user->id)->one();
                    $userel->setFieldValue('userCourse', $submissionUserCourseId);
                    $userel->setFieldValue('userProfileIntroduction',$submissionUserCourseId);
                    Craft::$app->elements->saveElement($userel);
                }
            }
        });
```

This semi works. The field 'userProfileIntroduction', which is a redactor field, gets filled in with the id.
![image](https://github.com/craftcms/cms/assets/46524074/a4a365e4-8ad2-48f5-9926-091986ded0e0)
The userCourse field remains empty.
![image](https://github.com/craftcms/cms/assets/46524074/9c80d71f-3d3b-4bfd-974f-3f7cde171f24)
Does anyone have any idea why this doesn't seem to work? As it is basically the same code that is used by the modify function.

**Comment by janhenckens** [comment]
Does the `userCourse` field get filled by Formie/Freeform on this submission? (I assume that's what you're using to build the form?)

If yes, then this line should return an element(query?)

```php
$submissionUserCourse = $submission->getFieldValue('userCourse')->siteId($primarySiteId)->one();
```

I'd try to debug the `submissionUserCourse` here, through xdebug or logging the result.

**Reply by JulieVelghe** [reply]
That line gets used by a profile form, to update the user data. That line propably gives an element query, but i'm using the id of that to update the field.
I would think that this code from after the edit (which works):
```
                        $submissionUserCourse = $submission->getFieldValue('userCourse')->siteId($primarySiteId)->one();
                        $submissionUserCourseId = $submissionUserCourse->id;
                        $user->setFieldValue('userCourse', $submissionUserCourseId);
```
would be the same as this code from after the creation (which doesn't work):
```
                        $submissionUserCourse = \craft\elements\Entry::find()->section('courses')->id($tempcourseid)->siteId($primarySiteId)->one();
                        $submissionUserCourseId = $submissionUserCourse->id;
                        $userel->setFieldValue('userCourse', $submissionUserCourseId);
```

**Comment by mmikkel** [comment, marked as answer]
When populating relational fields programmatically, you'll need to use an **array** of element IDs (even if it's just a single ID), i.e.

```php
$userel->setFieldValue('userCourse', [$submissionUserCourseId]);
```
